predicthq/kafka/consumer.py

```diff
@@ -0,0 +1,163 @@
+            topic = tp.topic
+            partition = tp.partition
+
+            if topic == self.input_topic:
+                if topic not in batch_ref:
+                    batch_ref[topic] = {}
+
+                for message in batch:
+                    # Skip null messages - they are used to 'delete' keys from
+                    # compacted topics.
+                    if message.value() is None:
+                        continue
+
+                    if message.error():
+                        error = message.error()
+                        if error.code() == confluent_kafka.KafkaError._PARTITION_EOF:
+                            continue
+                        else:
+                            raise confluent_kafka.KafkaException(msg.error())
+
+                    payload, ref = unpack_kafka_payload(message)
+                    message_id = ref['key']
+                    messages.append(Message(message_id, payload, ref))
+
+                batch_ref[topic][partition] = (batch[0].offset(), batch[-1].offset())
+
+            else:
+                log.warning('Received %(batch)s messages for unknown topic %(unknown_topic)s (expected %(expected_topic)s). Skipping.',
+                            {'batch': len(batch), 'unknown_topic': topic, 'expected_topics': self.input_topics})
+                continue
+
+        return messages, batch_ref
+
+    def _get_kafka_messages(self):
+        while True:
+            # The consumer interface doesn't obey the 'consumer_timeout_ms' consumer config.
+            log.debug('Calling consumer.consume, timeout %(time_s)s',
+                      {'time_s': self._consumer_timeout_ms})
+            messages = self._consumer.consume(num_messages=self._consumer_batch_size, timeout=self._consumer_timeout_ms / 1000)
+            if messages:
+                yield messages
+
+    def close(self):
+        if not self.closed:
+            self._consumer.close()
+            self.closed = True
+
+    def _process(self, producer=None):
+
+        log.info('Starting processor.')
+
+        messages_processed = 0
+
+        for kafka_messages in self._get_kafka_messages():
+            messages, batch_ref = self._format_message_batch(kafka_messages)
+            batch_size = len(messages)
+            start_ms = time.time() * 1000
+            log.info('[%(batch_ref)s] Starting processing batch of %(batch)d messages.',
+                     {'batch_ref': format_batch_ref(batch_ref), 'batch': batch_size})
+
+            try:
+                messages_to_produce = self.process_messages(messages)
+                if producer:
+                    producer.produce_batch(messages_to_produce)
+            except Exception as e:
+                log.exception('[%(batch_ref)s] Encountered exception while processing batch of messages.',
+                              {'batch_ref': format_batch_ref(batch_ref)})
+                raise
+            else:
+                self._consumer.commit()
```

Comment:
i knew you mentioned about not using autocommit but i think we need to have another discussion about this